{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:15.095927Z",
     "start_time": "2018-05-02T01:24:15.062468Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-01T18:25:05.087719Z",
     "start_time": "2018-05-01T18:25:05.054403Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.insert(\n",
    "    0,\n",
    "    '..',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:15.365464Z",
     "start_time": "2018-05-02T01:24:15.108661Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from sequencing_process.support.support.path import clean_path\n",
    "from sequencing_process.make_reference_genome import make_reference_genome\n",
    "from sequencing_process.download_clinvar_vcf_gz import download_clinvar_vcf_gz\n",
    "from sequencing_process.process_fastq_gz import check_fastq_gzs_using_fastqc, trim_fastq_gzs_using_skewer, align_fastq_gzs_using_bwa_mem\n",
    "from sequencing_process.process_bam import sort_and_index_bam_using_samtools_sort_and_index, mark_duplicates_in_bam_using_picard_markduplicates, check_bam_using_samtools_flagstat, get_variants_from_bam_using_freebayes_and_multiprocess, get_variants_from_bam_using_strelka\n",
    "from sequencing_process.process_vcf_gz import annotate_vcf_gz_using_snpeff, rename_chromosome_of_vcf_gz_using_bcftools_annotate, annotate_vcf_gz_using_bcftools_annotate, filter_vcf_gz_using_bcftools_view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:15.395486Z",
     "start_time": "2018-05-02T01:24:15.368440Z"
    }
   },
   "outputs": [],
   "source": [
    "DIRECTORY_PATH = clean_path('~/sequencing_process')\n",
    "\n",
    "assert os.path.isdir(DIRECTORY_PATH)\n",
    "\n",
    "OVERWRITE = True\n",
    "\n",
    "GRCH_DIRECTORY_PATH = clean_path('~/sequencing_process')\n",
    "\n",
    "assert os.path.isdir(GRCH_DIRECTORY_PATH)\n",
    "\n",
    "REGIONS = tuple('chr{}'.format(i) for i in range(1, 23)) + (\n",
    "    'chrX',\n",
    "    'chrY',\n",
    "    'chrM',\n",
    ")\n",
    "\n",
    "N_JOB = 1\n",
    "\n",
    "MEMORY = '8G'\n",
    "\n",
    "VARIANT_METHOD = 'freebayes'\n",
    "\n",
    "CLINVAR_VERSION = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:15.425202Z",
     "start_time": "2018-05-02T01:24:15.398135Z"
    }
   },
   "outputs": [],
   "source": [
    "try:\n",
    "\n",
    "    FASTA_GZ_FILE_PATH = make_reference_genome(GRCH_DIRECTORY_PATH)\n",
    "\n",
    "except FileExistsError:\n",
    "\n",
    "    FASTA_GZ_FILE_PATH = os.path.join(\n",
    "        GRCH_DIRECTORY_PATH,\n",
    "        'GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa.gz',\n",
    "    )\n",
    "\n",
    "    FASTA_FILE_PATH = os.path.splitext(FASTA_GZ_FILE_PATH)[0]\n",
    "\n",
    "    assert os.path.isfile(FASTA_FILE_PATH)\n",
    "\n",
    "try:\n",
    "\n",
    "    CLINVAR_VCF_GZ_FILE_PATH = download_clinvar_vcf_gz(\n",
    "        GRCH_DIRECTORY_PATH,\n",
    "        version=CLINVAR_VERSION,\n",
    "    )\n",
    "\n",
    "except FileExistsError:\n",
    "\n",
    "    CLINVAR_VCF_GZ_FILE_PATH = os.path.join(\n",
    "        GRCH_DIRECTORY_PATH,\n",
    "        [file_name for file_name in os.listdir(GRCH_DIRECTORY_PATH) if 'clinvar' in file_name and file_name.endswith('.gz')][0],\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:26.627411Z",
     "start_time": "2018-05-02T01:24:15.427956Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "skewer -x ../resource/general_bad_sequences.fasta -r 0 -d 0 --end-quality 30 --min 30 -n --output /home/k/sequencing_process/trimmed_fastq_gz/ --masked-output --excluded-output --threads 2 -m pe -y ../resource/general_bad_sequences.fasta /home/k/sequencing_process/simulation.bwa.read1.fastq.gz /home/k/sequencing_process/simulation.bwa.read2.fastq.gz\n",
      "/home/k/sequencing_process/trimmed_fastq_gz/trimmed.log:\n",
      "skewer v0.2.2 [April 4, 2016]\n",
      "COMMAND LINE:\tskewer -x ../resource/general_bad_sequences.fasta -r 0 -d 0 --end-quality 30 --min 30 -n --output /home/k/sequencing_process/trimmed_fastq_gz/ --masked-output --excluded-output --threads 2 -m pe -y ../resource/general_bad_sequences.fasta /home/k/sequencing_process/simulation.bwa.read1.fastq.gz /home/k/sequencing_process/simulation.bwa.read2.fastq.gz\n",
      "Input file:\t/home/k/sequencing_process/simulation.bwa.read1.fastq.gz\n",
      "Paired file:\t/home/k/sequencing_process/simulation.bwa.read2.fastq.gz\n",
      "trimmed:\t/home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair1.fastq, /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair2.fastq\n",
      "\n",
      "Parameters used:\n",
      "-- 3' end adapter sequences in file (-x):\t../resource/general_bad_sequences.fasta\n",
      "A:\tAGATCGGAAGAGC\n",
      "B:\tCTGTCTCTTATACACATCT\n",
      "-- paired 3' end adapter sequences in file (-y):\t../resource/general_bad_sequences.fasta\n",
      "01:\tAGATCGGAAGAGC\n",
      "02:\tCTGTCTCTTATACACATCT\n",
      "-- maximum error ratio allowed (-r):\t0.000\n",
      "-- maximum indel error ratio allowed (-d):\t0.000\n",
      "-- end quality threshold (-q):\t\t30\n",
      "-- minimum read length allowed after trimming (-l):\t30\n",
      "-- file format (-f):\t\tSanger/Illumina 1.8+ FASTQ (auto detected)\n",
      "-- number of concurrent threads (-t):\t2\n",
      "Mon Jul 16 12:55:27 2018 >> started\n",
      "\n",
      "Mon Jul 16 12:55:28 2018 >> done (1.003s)\n",
      "100000 read pairs processed; of these:\n",
      "     0 ( 0.00%) degenerative read pairs filtered out\n",
      "     0 ( 0.00%) short read pairs filtered out after trimming by size control\n",
      "100000 (100.00%) empty read pairs filtered out after trimming by size control\n",
      "     0 ( 0.00%) read pairs available.\n",
      "\n",
      "Length distribution of reads after trimming:\n",
      "length\tcount\tpercentage\n",
      "\n",
      "\n",
      "\n",
      "gzip --force /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair1.fastq\n",
      "gzip --force /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair2.fastq\n",
      "\n",
      "fastqc --threads 2 /home/k/sequencing_process/simulation.bwa.read1.fastq.gz /home/k/sequencing_process/simulation.bwa.read2.fastq.gz /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair1.fastq.gz /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "fastq_gz_0_file_path = os.path.join(\n",
    "    DIRECTORY_PATH,\n",
    "    'simulation.bwa.read1.fastq.gz',\n",
    ")\n",
    "\n",
    "assert os.path.isfile(fastq_gz_0_file_path)\n",
    "\n",
    "fastq_gz_1_file_path = os.path.join(\n",
    "    DIRECTORY_PATH,\n",
    "    'simulation.bwa.read2.fastq.gz',\n",
    ")\n",
    "\n",
    "assert os.path.isfile(fastq_gz_1_file_path)\n",
    "\n",
    "fastq_gz_0_trimmed_file_path, fastq_gz_1_trimmed_file_path = trim_fastq_gzs_using_skewer(\n",
    "    (\n",
    "        fastq_gz_0_file_path,\n",
    "        fastq_gz_1_file_path,\n",
    "    ),\n",
    "    output_directory_path=os.path.join(\n",
    "        DIRECTORY_PATH,\n",
    "        'trimmed_fastq_gz',\n",
    "    ),\n",
    "    n_job=N_JOB,\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "check_fastq_gzs_using_fastqc(\n",
    "    (\n",
    "        fastq_gz_0_file_path,\n",
    "        fastq_gz_1_file_path,\n",
    "        fastq_gz_0_trimmed_file_path,\n",
    "        fastq_gz_1_trimmed_file_path,\n",
    "    ),\n",
    "    n_job=N_JOB,\n",
    "    overwrite=OVERWRITE,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:36.898171Z",
     "start_time": "2018-05-02T01:24:26.631347Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "bwa index /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa.gz\n",
      "\n",
      "bwa mem -t 2 -v 3 /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa.gz /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair1.fastq.gz /home/k/sequencing_process/trimmed_fastq_gz/trimmed-pair2.fastq.gz | ../resource/k8-0.2.3/k8-linux ../resource/bwa-postalt.js /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa.gz.alt | samtools view -Sb --threads 2 > /home/k/sequencing_process/aligned.bam\n",
      "\n",
      "samtools sort --threads 2 /home/k/sequencing_process/aligned.bam > /home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam\n",
      "\n",
      "rm -rf /home/k/sequencing_process/aligned.bam\n",
      "\n",
      "samtools index -@ 2 /home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam\n",
      "\n",
      "picard -Xmx8G MarkDuplicates REMOVE_DUPLICATES=true INPUT=/home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam OUTPUT=/home/k/sequencing_process/duplicate_removed.bam METRICS_FILE=/home/k/sequencing_process/duplicate_removed.bam.metrics\n",
      "\n",
      "rm -rf /home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam\n",
      "\n",
      "rm -rf /home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam.bai\n",
      "/home/k/sequencing_process/duplicate_removed.bam.metrics:\n",
      "## htsjdk.samtools.metrics.StringHeader\n",
      "# MarkDuplicates INPUT=[/home/k/sequencing_process/sort_and_index_bam_using_samtools_sort_and_index.bam] OUTPUT=/home/k/sequencing_process/duplicate_removed.bam METRICS_FILE=/home/k/sequencing_process/duplicate_removed.bam.metrics REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "## htsjdk.samtools.metrics.StringHeader\n",
      "# Started on: Mon Jul 16 14:06:58 PDT 2018\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "samtools index -@ 2 /home/k/sequencing_process/duplicate_removed.bam\n",
      "\n",
      "samtools flagstat --threads 2 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/duplicate_removed.bam.flagstat\n",
      "/home/k/sequencing_process/duplicate_removed.bam.flagstat:\n",
      "0 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 mapped (N/A : N/A)\n",
      "0 + 0 paired in sequencing\n",
      "0 + 0 read1\n",
      "0 + 0 read2\n",
      "0 + 0 properly paired (N/A : N/A)\n",
      "0 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (N/A : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bam_file_path = align_fastq_gzs_using_bwa_mem(\n",
    "    (\n",
    "        fastq_gz_0_trimmed_file_path,\n",
    "        fastq_gz_1_trimmed_file_path,\n",
    "    ),\n",
    "    FASTA_GZ_FILE_PATH,\n",
    "    n_job=N_JOB,\n",
    "    output_bam_file_path=os.path.join(\n",
    "        DIRECTORY_PATH,\n",
    "        'aligned.bam',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "sorted_and_indexed_bam_file_path = sort_and_index_bam_using_samtools_sort_and_index(\n",
    "    bam_file_path,\n",
    "    remove_input_bam_file_path=True,\n",
    "    n_job=N_JOB,\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "duplicate_removed_bam_file_path = mark_duplicates_in_bam_using_picard_markduplicates(\n",
    "    sorted_and_indexed_bam_file_path,\n",
    "    memory=MEMORY,\n",
    "    remove_duplicates=True,\n",
    "    remove_input_bam_file_path_and_its_index=True,\n",
    "    n_job=N_JOB,\n",
    "    output_bam_file_path=os.path.join(\n",
    "        DIRECTORY_PATH,\n",
    "        'duplicate_removed.bam',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "check_bam_using_samtools_flagstat(\n",
    "    duplicate_removed_bam_file_path,\n",
    "    n_job=N_JOB,\n",
    "    overwrite=OVERWRITE,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:24:56.931056Z",
     "start_time": "2018-05-02T01:24:36.902214Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr1 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr5 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr6 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr2 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr7 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr3 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr8 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr9 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr4 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr10 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr13 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr11 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr14 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr12 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr15 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr17 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr16 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr18 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr21 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr22 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr19 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chrX /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chr20 /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chrM /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf.gz\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf.gz\n",
      "\n",
      "freebayes --fasta-reference /home/k/sequencing_process/GCA_000001405.15_GRCh38_full_plus_hs38DH-extra_analysis_set.fa --region chrY /home/k/sequencing_process/duplicate_removed.bam > /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf\n",
      "\n",
      "bgzip --threads 1 --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf && tabix --force /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf.gz\n",
      "\n",
      "bcftools concat --allow-overlaps --threads 2 /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf.gz /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf.gz > /home/k/sequencing_process/freebayes.vcf\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr1.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr2.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr3.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr4.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr5.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr6.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr7.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr8.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr9.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr10.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr11.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr12.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr13.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr14.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr15.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr16.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr17.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr18.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr19.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr20.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr21.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chr22.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrX.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrY.vcf.gz.tbi\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/get_variants_from_bam_using_freebayes.--region_chrM.vcf.gz.tbi\n",
      "\n",
      "bgzip --threads 2 --force /home/k/sequencing_process/freebayes.vcf && tabix --force /home/k/sequencing_process/freebayes.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "if VARIANT_METHOD not in (\n",
    "    'freebayes',\n",
    "    'strelka',\n",
    "):\n",
    "\n",
    "    raise ValueError('Unknown VARIANT_METHOD: {}.'.format(VARIANT_METHOD))\n",
    "\n",
    "if VARIANT_METHOD == 'freebayes':\n",
    "\n",
    "    vcf_gz_file_path = get_variants_from_bam_using_freebayes_and_multiprocess(\n",
    "        duplicate_removed_bam_file_path,\n",
    "        FASTA_FILE_PATH,\n",
    "        REGIONS,\n",
    "        n_job=N_JOB,\n",
    "        output_vcf_file_path=os.path.join(\n",
    "            DIRECTORY_PATH,\n",
    "            '{}.vcf'.format(VARIANT_METHOD),\n",
    "        ),\n",
    "        overwrite=OVERWRITE,\n",
    "    )\n",
    "\n",
    "    keep_filters = None\n",
    "\n",
    "    include_expression = '10<=DP & 30<=QUAL & 10<=(QUAL/AO) & 1<=SRF & 1<=SRR & 1<=SAF & 1<=SAR & 1<=RPR & 1<=RPL'\n",
    "\n",
    "elif VARIANT_METHOD == 'strelka':\n",
    "\n",
    "    vcf_gz_file_path = get_variants_from_bam_using_strelka(\n",
    "        duplicate_removed_bam_file_path,\n",
    "        FASTA_FILE_PATH,\n",
    "        os.path.join(\n",
    "            DIRECTORY_PATH,\n",
    "            'strelka',\n",
    "        ),\n",
    "        n_job=N_JOB,\n",
    "        overwrite=OVERWRITE,\n",
    "    )\n",
    "\n",
    "    keep_filters = ('PASS', )\n",
    "\n",
    "    include_expression = '30<=QUAL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-02T01:26:02.468212Z",
     "start_time": "2018-05-02T01:24:56.934994Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "bcftools view --regions chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,chr21,chr22,chrX,chrY,chrM --include '10<=DP & 30<=QUAL & 10<=(QUAL/AO) & 1<=SRF & 1<=SRR & 1<=SAF & 1<=SAR & 1<=RPR & 1<=RPL' --threads 2 /home/k/sequencing_process/freebayes.vcf.gz > /home/k/sequencing_process/filtered.vcf\n",
      "\n",
      "bgzip --threads 2 --force /home/k/sequencing_process/filtered.vcf && tabix --force /home/k/sequencing_process/filtered.vcf.gz\n",
      "\n",
      "bcftools annotate --rename-chrs ../resource/chrn_n.tsv --threads 2 /home/k/sequencing_process/filtered.vcf.gz > /home/k/sequencing_process/filtered.chromosome_renamed.vcf\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.vcf.gz.tbi\n",
      "\n",
      "bgzip --threads 2 --force /home/k/sequencing_process/filtered.chromosome_renamed.vcf && tabix --force /home/k/sequencing_process/filtered.chromosome_renamed.vcf.gz\n",
      "\n",
      "snpEff -Xmx8G -htmlStats /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.stats.html -csvStats /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.stats.csv -t -verbose -noLog GRCh38.86 /home/k/sequencing_process/filtered.chromosome_renamed.vcf.gz > /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.chromosome_renamed.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.chromosome_renamed.vcf.gz.tbi\n",
      "\n",
      "bgzip --threads 2 --force /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf && tabix --force /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.gz\n",
      "\n",
      "bcftools annotate --annotations /home/k/sequencing_process/clinvar.vcf.gz --threads 2 --columns =ID,INFO /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.gz > /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.clinvar.vcf\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.gz\n",
      "\n",
      "rm -rf /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.vcf.gz.tbi\n",
      "\n",
      "bgzip --threads 2 --force /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.clinvar.vcf && tabix --force /home/k/sequencing_process/filtered.chromosome_renamed.snpeff.clinvar.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "filtered_vcf_gz_file_path = filter_vcf_gz_using_bcftools_view(\n",
    "    vcf_gz_file_path,\n",
    "    regions=REGIONS,\n",
    "    keep_filters=keep_filters,\n",
    "    include_expression=include_expression,\n",
    "    n_job=N_JOB,\n",
    "    output_vcf_file_path=os.path.join(\n",
    "        DIRECTORY_PATH,\n",
    "        'filtered.vcf',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "chromosome_renamed_vcf_gz_file_path = rename_chromosome_of_vcf_gz_using_bcftools_annotate(\n",
    "    filtered_vcf_gz_file_path,\n",
    "    remove_input_vcf_gz_file_path_and_its_index=True,\n",
    "    n_job=N_JOB,\n",
    "    output_vcf_file_path=filtered_vcf_gz_file_path.replace(\n",
    "        '.vcf.gz',\n",
    "        '.chromosome_renamed.vcf',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "snpeff_annotated_vcf_gz_file_path = annotate_vcf_gz_using_snpeff(\n",
    "    chromosome_renamed_vcf_gz_file_path,\n",
    "    'GRCh38.86',\n",
    "    memory=MEMORY,\n",
    "    remove_input_vcf_gz_file_path_and_its_index=True,\n",
    "    n_job=N_JOB,\n",
    "    output_vcf_file_path=chromosome_renamed_vcf_gz_file_path.replace(\n",
    "        '.vcf.gz',\n",
    "        '.snpeff.vcf',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")\n",
    "\n",
    "clinvar_annotated_vcf_gz_file_path = annotate_vcf_gz_using_bcftools_annotate(\n",
    "    snpeff_annotated_vcf_gz_file_path,\n",
    "    CLINVAR_VCF_GZ_FILE_PATH,\n",
    "    ('--columns =ID,INFO', ),\n",
    "    remove_input_vcf_gz_file_path_and_its_index=True,\n",
    "    n_job=N_JOB,\n",
    "    output_vcf_file_path=snpeff_annotated_vcf_gz_file_path.replace(\n",
    "        '.vcf.gz',\n",
    "        '.clinvar.vcf',\n",
    "    ),\n",
    "    overwrite=OVERWRITE,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
